src/main/java/gov/nasa/jpf/constraints/solvers/ConstraintSolverService.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2015 United States Government, as represented by the Administrator
+ *                of the National Aeronautics and Space Administration. All Rights Reserved.
+ *           2017-2021 The jConstraints Authors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gov.nasa.jpf.constraints.solvers;
+
+import gov.nasa.jpf.constraints.api.ConstraintSolver;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.ServiceConfigurationError;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/** ServiceLoader for instantiating constraint solvers */
+public final class ConstraintSolverService {
+
+  private final ServiceLoader<ConstraintSolverProvider> loader;
+  private final Properties config = new Properties();
+  private final Map<String, ConstraintSolverProvider> providers =
+      new HashMap<String, ConstraintSolverProvider>();
+  private static final Logger logger = Logger.getLogger("constraints");
+  public static ConstraintSolverService service = new ConstraintSolverService();
```

Comment:
Since the service singleton is instantiated at class load time anyway, this is 100% equivalent to making everything static and not forcing clients to do `ConstraintSolverService.service.method()`